# This program obtimizes the vaules of parameters used in the magnet systems


The idea is to run the simulation to obtain polarization for many configurations and compre the polarization.<br>
The details of the simulation is explianed in simulation program.<br>
We can change this .ipynb file to .py file using:<br>
 jupyter nbconvert --to python Optimization.ipynb

In [1]:
import numpy as np
import math
import pandas as pd
import matplotlib.pyplot as plt
import random
import scipy.integrate as spi
from scipy.integrate import solve_ivp
import warnings

In [2]:
#suppress warnings
warnings.filterwarnings('ignore')

First define a function to obtain magnetic field

In [3]:
def ABC_Coil_CalculateB(x0,y0,z0,B2,B4,B5,B6):
    z=z0
    x=x0#0.55
    y=y0#0.55
    #zrange=zmax-zmin
    #z=np.arange(zmin,zmax,zrange/2500)
    #quadrupole and shim
    z1=63.6;   dz1=0.53;    r1=0.75;                # center,roll-off,radius [cm]
    zs=z1-5.25*2.54; zt=z1-.25*2.54; dzs=3.0;       # start,end,roll-off [cm]
    B1= 7500; Bs=B1*0.30;                           # field [G]  (Bs=0: no shim)
    vg=22000;
    # small,large ring
    z2=z1+7.0                      
    r2=10;          
    #B2=35;
    z3=z2;                          
    r3=15;          
    B3=-B2*pow((r2/r3),3)
    # top,bot of MSR
    za=z2+74.4;     zb=za+40.4;
    # small,large costheta
    z4=z2+8.4;      dz4=3.2;        r4=10;          #B4=60;
    z6=z2+60;       dz6=3.2;        r6=30;          #B6=2.5;
    #more parameters
    z5=250;                         r5=1.9;         #B5=0.03;
    Bx5=z*0+B5;
    Qxy=B1/r1/2*(1-np.tanh((z-z1)/dz1)) + Bs/r1/2*(np.tanh((z-zs)/dzs)-np.tanh((z-zt)/dzs));
    dQxy=-B1/r1/2*pow((1/np.cosh((z-z1)/dz1)),2)/dz1 + Bs/r1/2*(pow((1/np.cosh((z-zs)/dzs)),2) - pow(1/np.cosh((z-zt)/dzs),2))/dzs
    Bz2=B2*pow(r2,3)/pow((pow(r2,2)+pow((z-z2),2)),(3/2));
    Bz3=B3*pow(r3,3)/pow((pow(r3,2)+pow((z-z3),2)),(3/2));
    dBz2=-3*Bz2*(z-z2)/(pow(r2,2)+pow((z-z2),2));         #dBz/dz
    dBz3=-3*Bz3*(z-z3)/(pow(r3,2)+pow((z-z3),2));         #dBz/dz
    
    Bx4=B4*pow(r4,5)/pow((pow((pow(r4,2)+pow((z-z4),2)),(3/2)) + pow((pow(r4,2)-pow(dz4,2)),(3/2))),(5/3));
    Bx6=B6*pow(r6,5)/pow((pow((pow(r6,2)+pow((z-z6),2)),(3/2)) + pow((pow(r6,2)-pow(dz6,2)),(3/2))),(5/3));
    dBx4=-5*Bx4/(pow((pow(r4,2)+pow((z-z4),2)),(3/2)) + pow((pow(r4,2)-pow(dz4,2)),(3/2)))*np.sqrt(pow(r4,2)+ pow((z-z4),2))*(z-z4);
    dBx6=-5*Bx6/(pow((pow(r6,2)+pow((z-z6),2)),(3/2)) + pow((pow(r6,2)-pow(dz6,2)),(3/2)))*np.sqrt(pow(r6,2)+ pow((z-z6),2))*(z-z6);
    Bx = Qxy*y - 1/2*(dBz2+dBz3)*x + Bx4 + Bx6 + B5;
    By = Qxy*x - 1/2*(dBz2+dBz3)*y;
    Bz = dQxy*x*y + (Bz2+Bz3) + (dBx4+dBx6)*x;
    
    B  = [Bx,By,Bz];
    Bt=np.linalg.norm(B,axis=0);
    
    return B

Now defiene a function to obtain magnetic field in certain time

In [1]:
def Btm(x0,v,t):
    xt=[x0[0]+v[0]*t,x0[1]+v[1]*t,x0[2]+v[2]*t]
    B1tt=ABC_Coil_CalculateB(xt[0],xt[1],xt[2],B2,B4,B5,B6);
    return B1tt

In [2]:
def fA(t,y):
    return g*np.cross(y, Btm(x0,v,t,B2,B4,B5,B6))

In [3]:
def PolSim(z0,z1,B2,B4,B5,B6,nev):
    #B2=35;
    #B4=60;
    #B5=0.03;
    #B6=2.5;
    g=20378.9; v3=22000;
    #z0=62.5;
    r0=0.629;
    #z1=250;
    r1=5;
    #nev=10;
    div=(r1+r0)/(z1-z0);
    Pol=[];
    for n in range(1, nev):
        #variables
        rho=np.sqrt(random.random())*r0; 
        phi=random.random()*2*math.pi; # random position on entrance aperture
        x0=[ rho*np.cos(phi), rho*np.sin(phi), z0 ]; 
        B0=ABC_Coil_CalculateB(x0[0],x0[1],x0[2],B2,B4,B5,B6);
        #obtain norm of the magnetic field
        B0t=np.linalg.norm(B0,axis=0);
    
        r=99999; 
        while(r>r1):  # choose random direction within exit aperture
            ctheta=1-(1-np.cos(div))*random.random(); phi=random.random()*2*math.pi;
            v1=(v3*math.sqrt(1-pow(ctheta,2))*np.cos(phi), v3*math.sqrt(1-pow(ctheta,2))*np.sin(phi), v3*ctheta );
         
            x1=np.array(x0)
            v=np.array(v1)
            t1=(z1-z0)/v[2];
            x1=x0+(v*t1); 
            r=math.sqrt(x1[0]*x1[0] + x1[1]*x1[1]);
        
        B1  = ABC_Coil_CalculateB(x1[0],x1[1],x1[2],B2,B4,B5,B6);
        B1t = np.linalg.norm(B1,axis=0);
    
        #solve bloch equation using build-in function 
        sol = solve_ivp(lambda t, y: g*np.cross(y, Btm(x0,v,t,B2,B4,B5,B6)), [0,t1], B0/B0t, method="RK45", rtol = 1e-5)
        #sol = solve_ivp(fA, [0,t1], B0/B0t, method="RK45", rtol = 1e-5)
    
    
        #obtain last elemet of the array from the ode solution
        Mx=sol.y[0][-1]
        My=sol.y[1][-1]
        Mz=sol.y[2][-1]
        #norm of output vector of the ode
        Mn=np.sqrt(Mx**2+My**2+Mz**2)
    
        #polarization is given by 
        pol=(B1[0]*Mx+B1[1]*My+B1[2]*Mz)/(Mn*B1t)
        Pol.append(pol)
    return Pol

# Using Scipy Optimization modules to optimize the problem.
Online documentation:
https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.minimize.html

In [4]:
from scipy import optimize
from scipy.optimize import minimize
from scipy.optimize import dual_annealing

In [8]:
#define the optimization function:
def f3(B2_m):
    Polarization=PolSim(62.5,250,B2_m[0],B2_m[1],B2_m[2],B2_m[3],10)
    opt=(1.-(np.mean(Polarization)))
    #opt=B2_m[0]**2-12*B2_m[0]+20
    return opt

 Here we chose SLSQP method which stands for sequential least-square quadratic programming.<br>
 https://en.wikipedia.org/wiki/Sequential_quadratic_programming
 

In [9]:
B2_start=[35.,60,0.03,2.6];
B_bounds=[(20,40),(40,100),(0.01,1.0),(0,5)]
results = optimize.minimize(f3,B2_start,method='SLSQP',bounds=B_bounds)
#optimize.minimize(f3,[30.],method='SLSQP')
print('Status : %s' % results['message'])
print('Total Evaluations: %d' % results['nfev'])
solution = results['x']
evaluation = f3(solution)
print('Solution: f(%s) = %.5f' % (solution, evaluation))


Status : Optimization terminated successfully
Total Evaluations: 237
Solution: f([3.49350488e+01 6.00612003e+01 3.36806248e-02 2.58613590e+00]) = 0.00124


In [10]:
print(results)

     fun: 0.0005266829149174201
     jac: array([-12446.34804609, -22371.10326058, -13368.43076048, -37143.24088832])
 message: 'Optimization terminated successfully'
    nfev: 237
     nit: 18
    njev: 16
  status: 0
 success: True
       x: array([3.49350488e+01, 6.00612003e+01, 3.36806248e-02, 2.58613590e+00])


The Results we get from this is:<br>
Status : Optimization terminated successfully<br>
Total Evaluations: 237<br>
Solution: f([3.49350488e+01 6.00612003e+01 3.36806248e-02 2.58613590e+00]) = 0.00124 <br>
print(results)<br>
     fun: 0.0005266829149174201 <br>
     jac: array([-12446.34804609, -22371.10326058, -13368.43076048, -37143.24088832])<br>
 message: 'Optimization terminated successfully'<br>
    nfev: 237<br>
     nit: 18<br>
    njev: 16<br>
  status: 0<br>
 success: True<br>
       x: array([3.49350488e+01, 6.00612003e+01, 3.36806248e-02, 2.58613590e+00])<br>